### ISSUE TITLE
TimeSeries value quality information

### ISSUE THREAD

**Issue by MikeNeilson** [USACE/cwms-data-api#310]
It just dawned on me, we output the encoded quality integer, but from inside there API there is no documentation about the quality codes themselves nor an endpoint to retrieve them.

Not sure if just docs are sufficient, or maybe there should be an endpoint. or maybe a hosted javascript file to help with client side translation?

**Comment by krowvin** [comment]
The Excel file I have for this is very long. Can probably trim it back quite a bit and lookup a quality code conversion. At one point though i thought this was already being included in version 1 of the timeseries API? I could be wrong there. 

One thought, for the purpose of the API, would be to convert these values to sane human readable values. 

i.e. 0, 99, 2435, etc would return - OK, instead of their integers via a select on av_data_quality against the quality code. Or OK- Screened

At a bare minimum we could copy the CWMSWiki article on this and create a page in the documentation talking about this. The excel file is 348,162 lines. (32 bit values)

**Comment by krowvin** [comment]
UNKNOWN
OKAY
MISSING
QUESTIONABLE
REJECTED

Suggest, if we do the conversion from the code value to text that it be some combination of these from the spreadsheet. 
The values seem to repeat on these.

**Comment by MikeNeilson** [comment]
The values are encoded in various bitsets. So it's not a 1:1 mapping. (well, unless like the database schema you've already expanded the current options.)

**Comment by krowvin** [comment]
The view `av_data_quality` seems to, from the SQL in the spreadsheet on the second tab, have the same columns the spreadsheet has. 
Seen here:
```

select quality_code as unsigned_quality_code,
       case
       when quality_code <= 2147483647 then quality_code
       else quality_code - 4294967296
       end as signed_quality_code,
       screened_id,
       validity_id,
       range_id,
       changed_id,
       repl_cause_id,
       repl_method_id,
       test_failed_id,
       protection_id
  from av_data_quality
 order by 1
 ```
I would think that would allow us to do a cross lookup of the quality code for the text representation, then return something like this with the timeseries response:
``` 
"value-quality": {
0: "Unknown - Unscreened, 
3: "OKAY - Screened"
.. Return only the quality lookups that are in the result
},
"value-columns": [
    {
      "name": "string",
      "ordinal": 0,
      "datatype": "string"
    }
  ],
  "values": [
    [
      123,
      54.3,
      0
    ], 
    [
    456
    21.0,
    3
  ],
  ```
  
  But my reluctance is that this could potentially add to the query time. But it does serve as a means of a forced legend. It would also allow us to change these quality codes at some point if we ever needed to. (Changing what a 0 means, for example)
  
 Per your other suggestion, I could also create a minified javascript or json file for this that we could provide people to do clientside lookups of these quality codes. And assuming the file will never need to change it's okay if the file is on the client-side of things (A version number in/on the file would help)

**Comment by MikeNeilson** [comment]
Yeah, embedding it into the primary output will likely be an exercise in insanity. Better way to approach the design might be to think about how a, say "web based data validation editor" might use it (but like don't go so far as to actually make that editor.)

**Comment by adamkorynta** [comment]
![image](https://user-images.githubusercontent.com/47677856/216131218-ebb33d79-fad9-46cc-b19d-441744879b6c.png)

REGI has a tool for manipulating the USGS Rounding Spec. Something similar, but web hosted for quality would be neat.

**Comment by krowvin** [comment]
### I created a repo for this issue on my account
_We can migrate, if it would be better suited somewhere else_

We may still want to create a quality-code endpoint on CDA of some sort. This JS Library only takes into account the 32bit quality values stored in `AV_DATA_QUALITY`. 

It can be used to convert the decimal encoded value to the bit value or string representations _in the browser_ without additional AJAX calls.

1. Created an NPM package that wraps the `hec.data.tx` Java package as a JavaScript module. 

It allows you to do things like this, _on the client-side_
```
// JavaScript in the browser (or using NodeJS)
const qtx = new QualityTx()
qtx.getStringDescription(257, true) // Removing true returns it as it does in the original class
qtx.isScreened_int(257)
```
Which returns
```
>> {
    "CHANGED_ID": "ORIGINAL",
    "PROTECTION_ID": "UNPROTECTED",
    "QUALITY_CODE": "257",
    "RANGE_ID": "NO_RANGE",
    "REPL_CAUSE_ID": "AUTOMATIC",
    "REPL_METHOD_ID": "NONE",
    "SCREENED_ID": "SCREENED",
    "TEST_FAILED_ID": "NONE",
    "VALIDITY_ID": "UNKNOWN"
}

>> true
```
⚠ On every deploy a [test](https://github.com/krowvin/CWMS-Data-Quality/blob/master/tests/test.js) is ran against [this CSV file](https://github.com/krowvin/CWMS-Data-Quality/blob/master/tests/resources/quality_codes.csv) with over 380,000 lines to ensure they return correctly. 

It has most of the same methods the Java package has. 

NPM: [here](https://www.npmjs.com/package/cwms-data-quality)
GitHub: [here](https://github.com/krowvin/CWMS-Data-Quality)

2. I wrote an example page that uses the package here as a GitHub Page
https://krowvin.github.io/CWMS-Data-Quality/

This frontend source could be copied and adapted to any application. 
* The `Create` tab lets you enter bit map values and see their quality code  decimal values and binary representations. 
* The `Find` tab lets you enter decimal values and determine what they mean. 

Planning to convert to TypeScript with better JSDocs

**Comment by msweier** [comment]
Adding link for use case discussion at https://discourse.hecdev.net/t/best-practices-of-cwms-data-qualifer-codes/3805

**Comment by MikeNeilson** [comment]
So i don't want to do "computations" in CDA; however, extra "work" isn't unreasonable. Perhaps we should provide an option on time series retrieval to, instead of the int, return an array of text values instead so that some usages present the information directly?
